warp10/src/main/java/io/warp10/continuum/gts/MetadataTextComparator.java

```diff
@@ -24,158 +24,213 @@
     String name2 = o2.getName();
-    
+
     if (null == name1) {
       return -1;
     }
-    
+
     if (null == name2) {
       return 1;
     }
-    
+
     int comp = name1.compareTo(name2);
-    
+
     if (0 != comp) {
       return comp;
     }
-   
+
+    //
+    // Names are identical, compare labels
+    //
+
     if (0 == o1.getLabelsSize() && 0 == o2.getLabelsSize()) {
       return 0;
     }
-    
+
     if (0 == o1.getLabelsSize()) {
       return -1;
     }
-    
+
     if (0 == o2.getLabelsSize()) {
       return 1;
     }
-    
-    //
-    // Names are identical, compare labels
-    //
-    
+
     List<String> labels1 = new ArrayList<String>(o1.getLabelsSize());
     labels1.addAll(o1.getLabels().keySet());
-    
+
     List<String> labels2 = new ArrayList<String>(o2.getLabelsSize());
     labels2.addAll(o2.getLabels().keySet());
-    
+
     Collections.sort(labels1);
     Collections.sort(labels2);
-    
-    int idx1 = 0;
-    int idx2 = 0;
-    
-    while(idx1 < labels1.size() && idx2 < labels2.size()) {
-      
+
+    int idx = 0;
+
+    while (idx < labels1.size() && idx < labels2.size()) {
+
       // Compare label names
-      comp = labels1.get(idx1).compareTo(labels2.get(idx2));
-      
+      comp = labels1.get(idx).compareTo(labels2.get(idx));
+
       if (0 != comp) {
         return comp;
       }
-      
+
       // Compare label values
-      comp = o1.getLabels().get(labels1.get(idx1)).compareTo(o2.getLabels().get(labels2.get(idx2)));
-      
+      comp = o1.getLabels().get(labels1.get(idx)).compareTo(o2.getLabels().get(labels2.get(idx)));
+
       if (0 != comp) {
         return comp;
       }
-      
-      //
+
       // Advance labels
-      //
-      
-      idx1++;
-      idx2++;
+      idx++;
     }
 
-    if (idx1 == idx2) {
+    //
+    // Names are identical, labels are identical, compare attributes
+    //
+    
+    if (0 == o1.getAttributesSize() && 0 == o2.getAttributesSize()) {
       return 0;
     }
-    
-    if (idx1 < idx2) {
+
+    if (0 == o1.getAttributesSize()) {
       return -1;
-    } else {
+    }
+
+    if (0 == o2.getAttributesSize()) {
       return 1;
     }
+    
+
+    List<String> attr1 = new ArrayList<String>(o1.getAttributesSize());
+    attr1.addAll(o1.getAttributes().keySet());
+
+    List<String> attr2 = new ArrayList<String>(o2.getAttributesSize());
+    attr2.addAll(o2.getAttributes().keySet());
+
+    Collections.sort(attr1);
+    Collections.sort(attr2);
+
+    idx = 0;
+
+    while (idx < attr1.size() && idx < attr2.size()) {
+
+      // Compare attribute names
+      comp = attr1.get(idx).compareTo(attr2.get(idx));
+
+      if (0 != comp) {
+        return comp;
+      }
+
+      // Compare attribute values
+      comp = o1.getAttributes().get(attr1.get(idx)).compareTo(o2.getAttributes().get(attr2.get(idx)));
+
+      if (0 != comp) {
+        return comp;
+      }
+
+      // Advance attributes
+      idx++;
+    }
+
+    return 0;
```

Comment:
Ditto here, if we want to sort on attributes then GTS with fewer attributes should appear first

Reply:
Done